rocks-lib/src/project/rocks.rs

```diff
@@ -0,0 +1,221 @@
+//! Structs and utilities for `rocks.toml`
+
+use std::collections::HashMap;
+
+use itertools::Itertools as _;
+use serde::{Deserialize, Deserializer};
+use ssri::Integrity;
+use thiserror::Error;
+
+use crate::{
+    hash::HasIntegrity,
+    package::{PackageName, PackageReq, PackageVersion, PackageVersionReq},
+    rockspec::{
+        BuildSpec, ExternalDependencySpec, PerPlatform, PlatformSupport, RockDescription,
+        RockSource, Rocks, Rockspec, RockspecError, RockspecFormat, TestSpec,
+    },
+};
+
+#[derive(Clone, Debug, Deserialize)]
+pub struct RocksTomlInternal {
+    package: PackageName,
+    version: PackageVersion,
+    description: Option<RockDescription>,
+    supported_platforms: Option<PlatformSupport>,
+    #[serde(deserialize_with = "parse_map_to_package_vec")]
+    dependencies: PerPlatform<Vec<PackageReq>>,
+    #[serde(deserialize_with = "parse_map_to_package_vec_opt")]
+    build_dependencies: Option<PerPlatform<Vec<PackageReq>>>,
+    external_dependencies: Option<PerPlatform<HashMap<String, ExternalDependencySpec>>>,
+    #[serde(deserialize_with = "parse_map_to_package_vec_opt")]
+    test_dependencies: Option<PerPlatform<Vec<PackageReq>>>,
+    source: Option<PerPlatform<RockSource>>,
+    test: Option<PerPlatform<TestSpec>>,
+    build: PerPlatform<BuildSpec>,
+    hash: Integrity,
+}
+
+fn parse_map_to_package_vec<'de, D>(
+    deserializer: D,
+) -> Result<PerPlatform<Vec<PackageReq>>, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    let packages: PerPlatform<HashMap<PackageName, PackageVersionReq>> =
+        PerPlatform::deserialize(deserializer)?;
+
+    Ok(packages.map(|pkgs| {
+        pkgs.into_iter()
+            .map(|(name, version_req)| PackageReq { name, version_req })
+            .collect()
+    }))
+}
+
+fn parse_map_to_package_vec_opt<'de, D>(
+    deserializer: D,
+) -> Result<Option<PerPlatform<Vec<PackageReq>>>, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    let packages: Option<PerPlatform<HashMap<PackageName, PackageVersionReq>>> =
+        Option::deserialize(deserializer)?;
+
+    Ok(packages.map(|pkgs| {
+        pkgs.map(|pkgs| {
+            pkgs.into_iter()
+                .map(|(name, version_req)| PackageReq { name, version_req })
+                .collect()
+        })
+    }))
+}
+
+#[derive(Debug, Error)]
+pub enum IntoRockspecError {
+    #[error("no source url provided")]
+    NoSourceUrl,
```

Comment:
suggestion: Perhaps we could define a spec for generating a source URL for releases and scm versions, with variable expansion:

```toml
[source.scm]
url = "git://github.com/nvim-neorocks/$(PACKAGE)"

[source.release]
url = "https://github.com/nvim-neorocks/$(PACKAGE)/archive/v$(VERSION).zip"
```

For building a local project, we replace the `SourceSpec` with one that points to the project root.